src/Dialect/ONNX/ONNXOps.cpp

```diff
@@ -3172,41 +3172,83 @@ LogicalResult ONNXQuantizeLinearOp::inferShapes(
 
-  if (optionalAxis.has_value()) {
-    // axis attribute must be in the range [-r,r-1], where r = rank(input).
-    int64_t axis = optionalAxis.value();
-    if (axis < -xRank || axis >= xRank)
-      return onnx_mlir::Diagnostic::emitAttributeOutOfRangeError(
-          *this->getOperation(), "axis", axis,
-          onnx_mlir::Diagnostic::Range<int64_t>(-xRank, xRank - 1));
+    // TODO: Figure out whether to introduce a variant of this check from the
+    // spec ("'x_zero_point' and 'x' must have same type"). It is violated in
+    // in the resnet50-v1-12-qdq model where x, x_zero_point are i8, ui8.
+    //
+    // if (getElementType(x().getType()) != getElementType(zero.getType()))
+    //   return emitOpError("x and x_zero_point must have the same data type");
+
+    if (getElementType(zero.getType()).isInteger(32))
+      if (auto values = getDenseElementAttributeFromONNXValue(zero))
+        if (!values.isSplat() || !values.getSplatValue<APInt>().isZero())
+          return emitOpError("x_zero_point must be 0 for data type int32");
   }
 
-  // TODO: check other input constraints.
+  if (auto scaleTy = scale.getType().dyn_cast<RankedTensorType>()) {
+    if (scaleTy.getRank() > 1)
+      return emitOpError("x_scale must be a scalar or 1-D tensor");
+
+    // If x_scale is a scalar then quantization is per-tensor / per layer.
+    if (isScalar(scaleTy))
+      return success(); // Ignore axis. Nothing more to verify.
+
+    if (!isNonScalar1D(scaleTy))
+      return success(); // Cannot verify more until known if x_scale is scalar.
+
+    // If x_scale is a non-scalar 1-D tensor then quantization is per-axis.
+    if (auto xTy = x().getType().dyn_cast<RankedTensorType>()) {
+      int64_t r = xTy.getRank();
+      // axis attribute must be in the range [-r,r-1].
+      int64_t a = axis();
+      if (a < -r || a >= r)
+        return onnx_mlir::Diagnostic::emitAttributeOutOfRangeError(
+            *this->getOperation(), "axis", a,
+            onnx_mlir::Diagnostic::Range<int64_t>(-r, r - 1));
+      if (a < 0)
+        a += r;
+      if (!xTy.isDynamicDim(a) && xTy.getDimSize(a) != scaleTy.getDimSize(0))
+        return emitOpError(
+            "x_scale 1-D tensor length must match the input axis dim size");
+    } else {
+      // Cannot verify more until x rank is known.
+    }
+  } else {
+    // Cannot verify more until x_scale rank is known.
+  }
 
   return success();
 }
 
 LogicalResult ONNXDequantizeLinearOp::inferShapes(
     std::function<void(mlir::Region &)> doShapeInference) {
-  auto inTy = x().getType().dyn_cast<RankedTensorType>();
-  if (!inTy) {
-    return success();
-  }
-
-  auto yTy = y().getType().cast<ShapedType>();
-
-  if (!yTy.hasStaticShape()) {
-    FloatType f32 = FloatType::getF32(getContext());
-    RankedTensorType outType = RankedTensorType::get(inTy.getShape(), f32);
-    y().setType(outType);
-  }
+  if (auto xTy = x().getType().dyn_cast<RankedTensorType>())
+    updateType(y(), xTy.getShape());
```

Comment:
No need to add. It seems possible to me that  the shape[0] of zero() or scale() can be used to refine X and Y when X and Y are unranked and zero or scale is 1D. It's a corner case.

Reply:
> It seems possible to me that the shape[0] of zero() or scale() can be used to refine X and Y when X and Y are unranked and zero or scale is 1D.

good idea, I implemented this

I also added one more check in verify, namely x_zero_point non-scale 1-D rank should match input axis dim size

please take another look